beps/0002-dynamic-frontend-plugins/README.md

```diff
@@ -0,0 +1,492 @@
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this BEP. Describe why the change is important and the benefits to users.
+-->
+
+### Goals
+
+<!--
+List the specific goals of the BEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- discover and choose tooling to enable dynamic frontend plugins
+- easy way of converting existing frontend plugins to be dynamic
+- turning on and off UI plugins at runtime without the requirement of rebuilding or restarting backstage
+- manage dynamic plugins declaratively
+
+An additional goal might be exploring the scope of a "on demand" rending and loading. Currently all plugins, or at least some portion of them must be loaded at UI bootstrap. This approach is not fully dynamic and might be considered a waste of resources. Fully optimizing dynamic plugins loading would require additional changes to the new UI system.
+
+### Non-Goals
+
+<!--
+What is out of scope for this BEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation.
+-->
+
+### Definition of UI dynamic plugin
+
+A dynamic UI plugin (from now just plugin) is a plugin that is not part of the output of a backstage instance build. The plugin and its assets are injected into backstage at runtime. In this case, its injected into the browser at some point during user session.
+
+From the user POV, there is no difference between classic and dynamic plugins.
+
+The difference is known only to maintainers and should be limited to
+
+- build requirements
+- integration into backstage
+
+### Dynamic loading tool
+
+Traditional plugins are not dynamic out of the box. Additional changes are required during build time to make a plugin dynamic.
+
+The main application (shell) also requires some changes to be able to load, inject, and propagate mandatory context to dynamic plugins.
+
+Over the last few year, [Module Federation](https://github.com/module-federation) has become the standard when it comes to dynamically load JS modules browser (and nodejs) environment from remote locations and sharing context between the shell and the remote module.
+
+Although there are other options, like externalizing dependencies, the module federation has proven itself as a robust solution to this particular problem.
+
+### Module federation implementation
+
+There are multiple available implementations of module federation.
+
+Historically, module federation was implemented as a [Webpack feature](https://webpack.js.org/). Since then, additional implementations were created. Mainly for [Rspack](https://www.rspack.dev/) and [Vite](https://github.com/originjs/vite-plugin-federation).
```

Comment:
```suggestion
Historically, module federation was implemented as a [Webpack feature](https://webpack.js.org/). Since then, additional implementations were created. Mainly for [`rspack`](https://www.rspack.dev/) and [Vite](https://github.com/originjs/vite-plugin-federation).
```